flow/layers/layer.h

```diff
@@ -122,6 +122,10 @@ class Layer {
     const RasterCache* raster_cache;
     const bool checkerboard_offscreen_layers;
     const float frame_device_pixel_ratio;
+
+    bool needs_painting(const SkRect& paint_bounds) {
+      return !internal_nodes_canvas->quickReject(paint_bounds);
```

Comment:
What's the difference between the `paint_bounds` argument, and the `paint_bounds_` member variable? I thought they should agree with each other?

Reply:
This method was on PaintContext so it needed to have it passed in.

I've since moved the method back to Layer so it is now "needs_painting(PaintContext& context);"

This code is obsolete.

Reply:
New code that now has `Layer::needs_painting(PaintContext& context)` pushed.

The new code also has `Layer::is_empty()` because there were 2 or 3 places that wanted to ask if the layer (usually the root) needed to be painted but they hadn't constructed a full PaintContext yet. So, those places will now depend solely on whether the layer is empty.